### PR TITLE
ARXIVNG-293: mypy typing and domain dataclasses

### PR DESCRIPTION
**Pull request by JaimieMurdock**

### CODE REVIEW

File: search/domain/base.py [hunk 1]
```diff
@@ -1,147 +1,84 @@
...
 
+def asdict(obj: Any) -> dict:
+    """Coerce a dataclass object to a dict."""
+    return {key: value for key, value in _asdict(obj).items() if value}
 
-class DocMeta(SchemaBase):
-    """Metadata for an arXiv paper, retrieved from the core repository."""
 
+@dataclass
+class DocMeta:
```
Comment (JaimieMurdock) [inline search/domain/base.py]: Wow. Great work on getting all these defaults set!

File: search/domain/base.py [hunk 2]
```diff
@@ -1,147 +1,84 @@
...
 
-class DateRange(Base):
-    """Represents an open or closed date range."""
 
-    start_date = Property('start_date', datetime)
+@dataclass
+class DateRange:
+    """Represents an open or closed date range, for use in :class:`.Query`."""
+
+    start_date: datetime = datetime(1990, 1, 1, tzinfo=EASTERN)
```
Comment (JaimieMurdock) [inline search/domain/base.py]: This was arbitrarily selected.

File: search/domain/base.py [hunk 3]
```diff
@@ -171,46 +109,113 @@ def __str__(self):
...
+@dataclass
+class SimpleQuery(Query):
+    """Represents a simple search query."""
+
+    field: str = ''
+    value: str = ''
+
+
+@dataclass(init=True)
+class Document:
```
Comment (JaimieMurdock) [inline search/domain/base.py]: Again, wow on the defaults here.

**Comment by mhl10** [comment]
Looks good to me, though I defer approval to @erickpeirson despite conflict of interest ;-)

File: tests/test_transform.py [hunk 4]
```diff
@@ -71,22 +74,22 @@ def test_owners(self):
 
     def test_submitted_date(self):
         """Field ``submitted_date`` is populated from ``submitted_date``."""
-        meta = {
+        meta = DocMeta(**{
```
Comment (JaimieMurdock) [inline tests/test_transform.py]: Nice use of `**` to convert!

**Pull request merged**